### PR TITLE
Add environment tab to run configuration (again) and wire to CloudSdk

### LINKED ISSUE #429: Allow easy way to provide environment variables for Cloud emulators
Local emulators for  gcloud-java (https://github.com/GoogleCloudPlatform/gcloud-java)  for datastore  and pubsub sets environment variable in the current shell.

that is, if i run
gcloud beta emulators datastore start
it will set
export DATASTORE_LOCAL_HOST=localhost:8087
and the gcloud-java datastore client will use that to connect locally.

Its not clear how to set that variable in the eclipse tools (i tried setting it in the launch config manually but also saw the following:
launch_config_tabs

Bonus to find some automagic or easy way to read in these variables without manually adding them in to the runtime every time.   NOTE: I dont' think the dev_appserver allows for dynamic env variables (you need to declare them directly in appengine-web.xml or app.yaml
see:
https://cloud.google.com/appengine/docs/java/config/appref

### PR DESCRIPTION
**Pull request by briandealwis**
Provides a stop-gap measure for specifying Google services.

Fixes #429 and allows specifying `GOOGLE_APPLICATION_CREDENTIALS` and `GOOGLE_CLOUD_PROJECT` as per #2352.

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java [hunk 1]
```diff
@@ -294,18 +301,48 @@ public void testGenerateRunConfiguration_withProgramArgs() throws CoreException
...
+    assertEquals("bar", parsedEnvironment.get("foo"));
+    assertEquals(System.getenv("PATH"), parsedEnvironment.get("baz"));
+    verify(launchConfiguration).getAttribute(eq(ILaunchManager.ATTR_ENVIRONMENT_VARIABLES),
+        anyMapOf(String.class, String.class));
+    verify(launchConfiguration, atLeastOnce())
+        .getAttribute(eq(ILaunchManager.ATTR_APPEND_ENVIRONMENT_VARIABLES),
+        anyBoolean());
+  }
+
+  @Test(expected = CoreException.class)
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java]: Nit: `expected` is discouraged per internal doc. Use try-catch with `fail()`.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java [hunk 2]
```diff
@@ -299,6 +299,22 @@ DefaultRunConfiguration generateServerRunConfiguration(ILaunchConfiguration conf
...
 
+    boolean environmentAppend =
+        configuration.getAttribute(ILaunchManager.ATTR_APPEND_ENVIRONMENT_VARIABLES, true);
+    if (!environmentAppend) {
+      throw new CoreException(StatusUtil.error(this, "'Replace environment' not yet supported"));
+    }
+    // getEnvironment() applies var substitutions, but also then joins the key-value pairs
+    String[] environment = getEnvironment(configuration);
+    if (environment != null) {
+      Map<String, String> asMap = new HashMap<String, String>();
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: Nit: `new HashMap<>();`

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java [hunk 3]
```diff
@@ -299,6 +299,22 @@ DefaultRunConfiguration generateServerRunConfiguration(ILaunchConfiguration conf
       devServerRunConfiguration.setAdditionalArguments(programArguments);
     }
 
+    boolean environmentAppend =
+        configuration.getAttribute(ILaunchManager.ATTR_APPEND_ENVIRONMENT_VARIABLES, true);
+    if (!environmentAppend) {
+      throw new CoreException(StatusUtil.error(this, "'Replace environment' not yet supported"));
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: Can the error message more specific to point out that the _Replace environment_ is a checkbox in the _Environment_ launch tab? Users will hit the error far later when actually executing the dev appserver, so users can get easily lost about what this means.

Ideally, the _Replace environment_ checkbox should be disabled, but I guess this is difficult to implement?

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java [hunk 4]
```diff
@@ -294,18 +301,48 @@ public void testGenerateRunConfiguration_withProgramArgs() throws CoreException
...
+    DefaultRunConfiguration config = new LocalAppEngineServerLaunchConfigurationDelegate()
+        .generateServerRunConfiguration(launchConfiguration, server, ILaunchManager.RUN_MODE);
+
+    Map<String, String> parsedEnvironment = config.getEnvironment();
+    assertNotNull(parsedEnvironment);
+    assertEquals("bar", parsedEnvironment.get("foo"));
+    assertEquals(System.getenv("PATH"), parsedEnvironment.get("baz"));
+    verify(launchConfiguration).getAttribute(eq(ILaunchManager.ATTR_ENVIRONMENT_VARIABLES),
+        anyMapOf(String.class, String.class));
+    verify(launchConfiguration, atLeastOnce())
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java]: I don't see much value in these `verify` checks. We already check the values of `parsedEnvironment`.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java (hunk 3 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: Unlike some of the other tabs, it is possible but it's brittle (fragile superclass and all).  I forgot to open an issue with appengine-plugins-core to add support for append/replace (https://github.com/GoogleCloudPlatform/appengine-plugins-core/issues/446).

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java (hunk 4 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegateTest.java]: We've been using them pretty systematically throughout.

**Comment by briandealwis** [comment]
`AbstractLaunchConfigurationDelegate#getEnvironment()` checks and does the append-to-existing environment as part of its handling, which appengine-plugins-core does too.  I think we're better off deferring to appengine-plugins-core.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java (hunk 3 above)
Comment (chanseokoh) [reply plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: > Can the error message be more specific to point out that the _Replace environment_ is a checkbox in the _Environment_ launch tab?

Can we still do this? I think the current error message is too distance from the actual error source.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java [hunk 5]
```diff
@@ -299,6 +302,25 @@ DefaultRunConfiguration generateServerRunConfiguration(ILaunchConfiguration conf
...
+    }
+    // Could use getEnvironment(), but it does `append` processing and joins as key-value pairs
+    Map<String, String> environment = configuration.getAttribute(
+        ILaunchManager.ATTR_ENVIRONMENT_VARIABLES, Collections.<String, String>emptyMap());
+    if (environment != null) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: Probably we don't need this null check, as we return the empty set when it's undefined.

File: plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java (hunk 3 above)
Comment (briandealwis) [reply plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEngineServerLaunchConfigurationDelegate.java]: oops, will do

**Comment by chanseokoh** [comment]
LGTM. BTW, in my Eclipse testing, the Append/Replace environment options are disabled. They are enabled in your environment?

selection_003

**Comment by briandealwis** [comment]
It's disabled until a variable has been added:

screen shot 2017-09-11 at 3 18 35 pm

screen shot 2017-09-11 at 3 17 58 pm

**Comment by briandealwis** [comment]
@chanseokoh can you tick the approve box when you have a moment?

**Pull request merged**